### PR TITLE
Updates div to perform true division

### PR DESCRIPTION
**Pull request by mruberry**
This PR:

- updates div to perform true division
- makes torch.true_divide an alias of torch.div

This follows on work in previous PyTorch releases that first deprecated div performing "integer" or "floor" division, then prevented it by throwing a runtime error.

### CODE REVIEW

**Comment by mruberry** [comment]
@ailzhang Looks like this change is breaking XLA because it changes `torch.true_divide` to be an alias for `torch.div`. It will also change the behavior of `torch.div` to always perform true division.

**Comment by ailzhang** [comment]
cc: @JackCaoG This requires updating `torch.div` behavior on xla side, would you mind taking a look?

**Comment by mruberry** [comment]
> cc: @JackCaoG This requires updating `torch.div` behavior on xla side, would you mind taking a look?

ping @JackCaoG. We'd like to get this fix in sooner rather than later.

**Comment by JackCaoG** [comment]
@mruberry Sorry, missed this email. I will take a look tomorrow. Feel free to ping me on Slack if I didn't response to these kind of failures within a day 😄 .

**Comment by mruberry** [comment]
> @mruberry Sorry, missed this email. I will take a look tomorrow. Feel free to ping me on Slack if I didn't response to these kind of failures within a day 😄 .

haha, thanks @JackCaoG, but it's not "next day" urgent! It's like "by next week" urgent, if that works for you? Should be a small fix.

**Comment by JackCaoG** [comment]
@mruberry Sounds good, I will work on it tomorrow.

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -106,8 +106,15 @@ def rsub(g, self, other, alpha=None):
...
-
+# Division in PyTorch is always "true" division like Python 3
+# Trying to mimick this in ONNX would error-prone, so this throws a warning
+#   telling users how to structure their networks, instead.
 def div(g, self, other):
+    warnings.warn("Div in ONNX performs floor division with integer inputs, " +
+                  "unlike division in PyTorch which always performs true " +
+                  "division. Ensure your inputs to div are the same dtype " +
+                  "and cast them to a float dtype before performing the " +
+                  "division to the same behavior as PyTorch.")
```
Comment (mruberry) [inline torch/onnx/symbolic_opset9.py:117]: wordsmith

**Comment by JackCaoG** [comment]
Hey @mruberry I am running into an issue where if I do 
```
import torch, torch_xla                                                                                                                                                                                                                                     
import torch_xla.core.xla_model as xm                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                            
device = xm.xla_device()                                                                                                                                                                                                                                    
t1 = torch.tensor(10, device=device)                                                                                                                                                                                                                        
print(torch.div(t1, 2.0))    
```

the two tensor version of the divide `at::Tensor AtenXlaType::div(const at::Tensor& self, const at::Tensor& other)` will be called instead of the sclar version `at::Tensor AtenXlaType::div(const at::Tensor& self, at::Scalar other)` . `const at::Tensor& other` would have the dtype of `float64` instead of the `float32`.  Is this expected?

**Comment by mruberry** [comment]
> Hey @mruberry I am running into an issue where if I do
> 
> ```
> import torch, torch_xla                                                                                                                                                                                                                                     
> import torch_xla.core.xla_model as xm                                                                                                                                                                                                                       
>                                                                                                                                                                                                                                                             
> device = xm.xla_device()                                                                                                                                                                                                                                    
> t1 = torch.tensor(10, device=device)                                                                                                                                                                                                                        
> print(torch.div(t1, 2.0))    
> ```
> 
> the two tensor version of the divide `at::Tensor AtenXlaType::div(const at::Tensor& self, const at::Tensor& other)` will be called instead of the sclar version `at::Tensor AtenXlaType::div(const at::Tensor& self, at::Scalar other)` . `const at::Tensor& other` would have the dtype of `float64` instead of the `float32`. Is this expected?

No it's not. Let me do some local testing. I thought we had tests for this. I also need to rebase this.

**Comment by mruberry** [comment]
@JackCaoG I believe I fixed the issue you were seeing. This PR is also rebased - sorry there was a holdup.

File: aten/src/ATen/native/PointwiseOps.cpp [hunk 2]
```diff
@@ -73,11 +73,12 @@ Tensor& addcdiv_out(
...
-      "for integral inputs (self + value * tensor1 // tensor2) and ",
-      "division for float inputs (self + value * tensor1 / tensor2). ",
-      "The future addcdiv behavior can be implemented with true_divide ",
-      "(self + value * torch.true_divide(tensor1, tensor2).");
+      "The historic addcdiv behavior can be implemented as ",
+      "(input + value * torch.trunc(tensor1 / tensor2)).to(input.dtype) ",
+      "for integer inputs and as ",
+      "(input + value * tensor1 / tensor2) for float inputs. ",
+      "The future addcdiv behavior is just the latter implementation: ",
+      "(input + value * tensor1 / tensor2).");
```
Comment (mruberry) [inline aten/src/ATen/native/PointwiseOps.cpp:81]: , for all dtypes.

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -311,24 +311,20 @@ def merge_dicts(*dicts):
...
-    The future addcdiv behavior can be implemented with :func:`true_divide`
-    (:attr:`input` + :attr:`value` * torch.true_divide(:attr:`tensor1`,
-    :attr:`tensor2`).
+    Integer division with addcdiv is no longer supported, and in a future
+    release addcdiv will perform a true division of tensor1 and tensor2.
+    The historic addcdiv behavior can be implemented as
+    (input + value * torch.trunc(tensor1 / tensor2)).to(input.dtype)
+    for integer inputs and as (input + value * tensor1 / tensor2) for float inputs.
+    The future addcdiv behavior is just the latter implementation:
+    (input + value * tensor1 / tensor2).
```
Comment (mruberry) [inline torch/_torch_docs.py:327]: This note also needs the update above

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1226,6 +1226,34 @@
   use_c10_dispatcher: full
   variants: method
 
+  # true_divide, an alias for div
+- func: true_divide.Tensor(Tensor self, Tensor other) -> Tensor
+  use_c10_dispatcher: full
+  variants: function, method
+  dispatch:
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:1233]: Does setting dispatch override the gradient? Check what test would catch this.

Delete dispatches.

Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:1233]: Interesting follow-up: probably want to add gradcheck to test_op_aliases, method_test entries missing for true_divide and floor_divide.

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -2371,68 +2367,37 @@ def merge_dicts(*dicts):
     tensor(2.6537)
 """.format(**common_args))
 
-add_docstr(torch.div,
-           r"""
+add_docstr(torch.div, r"""
 div(input, other, *, out=None) -> Tensor
```
Comment (mruberry) [inline torch/_torch_docs.py:2371]: Might be nice to start adding types here

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -2371,68 +2367,37 @@ def merge_dicts(*dicts):
...
+
+Supports :ref:`broadcasting to a common shape <broadcasting-semantics>`,
+:ref:`type promotion <type-promotion-doc>`, and integer, float, and complex inputs.
+Always promotes integer types to the default scalar type.
 
 Args:
-    {input}
-    other (Number): the number to be divided to each element of ``input``
+    input (Tensor): the dividend
+    other (Tensor or Scalar): the divisor
```
Comment (mruberry) [inline torch/_torch_docs.py:2389]: Tensor or Number

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -2371,68 +2367,37 @@ def merge_dicts(*dicts):
...
-:class:`torch.dtype` of the result tensor is determined following rules
-described in the type promotion :ref:`documentation <type-promotion-doc>`. If
-``out`` is specified, the result must be :ref:`castable <type-promotion-doc>`
-to the :class:`torch.dtype` of the specified output tensor. Integral division
-by zero leads to undefined behavior.
+.. note::
+    Performs a "true" division like Python 3. See :func:`torch.floor_divide`
+    for floor division.
+
+Supports :ref:`broadcasting to a common shape <broadcasting-semantics>`,
```
Comment (mruberry) [inline torch/_torch_docs.py:2383]: TODO: check if scalar semantics are covered

**Comment by JackCaoG** [comment]
@mruberry I am still seeing the same behavior, running
```
import torch, torch_xla                                                                                                                                                                                                         
import torch_xla.core.xla_model as xm                                                                                                                                                                                           
                                                                                                                                                                                                                                
device = xm.xla_device()                                                                                                                                                                                                        
t1 = torch.tensor(10, device=device)                                                                                                                                                                                            
print(torch.div(t1, 2.0))   
```
give me

```
div two tensor
div two tensor
tensor(5., device='xla:0', dtype=torch.float64)
```

I am using the most recent div_true and I also rebased my pt/xla branch
```
(pytorch2) jackcao@jackcao64:~/Desktop/gitProj2/pytorch$ git status
On branch div_true
Your branch is up to date with 'origin/div_true'.
```

Could you give it another look?

**Comment by mruberry** [comment]
@JackCaoG Interesting...

So true_divide is just calling div:

```
Tensor true_divide(const Tensor& self, Scalar divisor) {
  return self.div(divisor);
}
```

Which wraps the scalar and redispatches to div:

https://github.com/pytorch/pytorch/blob/9f54bcc522747aed1377a9d56a6511d7e2b38daf/aten/src/ATen/native/BinaryOps.cpp#L406

The code for div itself isn't changing in this PR. What was the behavior before this PR for div?

**Comment by JackCaoG** [comment]
@mruberry I am using the pytorch head against the pytorch/xla head to run the same code and I am seeing the same behavior.

```
torch.div(t1, 2.0)
```
will call `AtenXlaType::div(const at::Tensor& self, const at::Tensor& other)` instead of `AtenXlaType::div(const at::Tensor& self, at::Scalar other)`. This behavior is unrelated to this pr, but it does seems wrong.

**Comment by mruberry** [comment]
> @mruberry I am using the pytorch head against the pytorch/xla head to run the same code and I am seeing the same behavior.
> 
> ```
> torch.div(t1, 2.0)
> ```
> 
> will call `AtenXlaType::div(const at::Tensor& self, const at::Tensor& other)` instead of `AtenXlaType::div(const at::Tensor& self, at::Scalar other)`. This behavior is unrelated to this pr, but it does seems wrong.

I agree with you but I'm glad we've verified it's not a regression. Want to file a follow-up issue on the PyTorch/XLA side? 

How does this PR look otherwise?

**Comment by JackCaoG** [comment]
> > @mruberry I am using the pytorch head against the pytorch/xla head to run the same code and I am seeing the same behavior.
> > ```
> > torch.div(t1, 2.0)
> > ```
> > 
> > 
> > will call `AtenXlaType::div(const at::Tensor& self, const at::Tensor& other)` instead of `AtenXlaType::div(const at::Tensor& self, at::Scalar other)`. This behavior is unrelated to this pr, but it does seems wrong.
> 
> I agree with you but I'm glad we've verified it's not a regression. Want to file a follow-up issue on the PyTorch/XLA side?
> 
> How does this PR look otherwise?

This pr looks good. I need to fix/workaround the type promotion issue caused by the `scalar->tensor` call.

File: torch/_torch_docs.py (hunk 7 above)
Comment (mruberry) [reply torch/_torch_docs.py:2383]: The broadcasting note DOES NOT describe this, and the note is in need of an udpate. I added a note to our Docs Hackathon planner.

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:1233]: Explicitly dispatching was causing autograd to fail. Deleting the dispatches works. Unfortunately aliases can't be added to method_tests() without additional test changes because test_jit.py has a check where it looks for the node, by name, but the node doesn't appear in the JIT's graph (because it's been aliased to another op). In the near future aliases will be refactored into OpInfos and gradcheck'd automatically there.

File: aten/src/ATen/native/TensorIterator.cpp [hunk 8]
```diff
@@ -816,7 +816,7 @@ void TensorIterator::select_all_keeping_dim(int start_dim, IntArrayRef indices)
 }
 
 TensorIterator TensorIterator::binary_op(Tensor& out, const Tensor& a,
```
Comment (mruberry) [inline aten/src/ATen/native/TensorIterator.cpp:818]: binary_op_float

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -410,11 +410,15 @@ def method_tests():
         ('div', (S, S, S), (uniform_scalar(0.1),), 'scalar_broadcast_rhs', (True,)),
         ('div', (), (uniform_scalar(0.1),), 'scalar_broadcast_lhs', (True,)),
         ('div', torch.rand(S, S, S) + 1e-1, (3.14,), 'constant', (True,)),
+        ('div', uniform_scalar(1e-1, requires_grad=True), (3.14,), 'scalar_constant', (True,)),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:413]: add true_divide and ignore in jit (for now) with comment

File: test/test_type_promotion.py [hunk 10]
```diff
@@ -590,58 +590,61 @@ def test_transpose(self, device):
...
+
+            # Tests that requests for an integer quotient fail
+            if not dtype.is_floating_point:
+                with self.assertRaises(RuntimeError):
+                    op(dividend, divisor)
+                with self.assertRaises(RuntimeError):
+                    op(dividend, 2)
+            else:
+                # Tests that requests for a floating quotient succeed
+                div_result = dividend.clone().div_(divisor)
```
Comment (gchanan) [inline test/test_type_promotion.py:645]: should this use op?

File: test/test_type_promotion.py [hunk 11]
```diff
@@ -590,58 +590,61 @@ def test_transpose(self, device):
...
-                dividend.true_divide_(divisor)
-            with self.assertRaises(RuntimeError):
-                dividend.true_divide_(2)
-        else:
-            # Tests that requests for a floating quotient succeed
-            div_result = dividend.clone().div_(divisor)
-            self.assertEqual(div_result, dividend.clone().true_divide_(divisor))
-            self.assertEqual(dividend.clone().div_(2), dividend.clone().true_divide_(2))
+    def test_div_promotion_inplace(self, device, dtype):
+        for op in (torch.Tensor.div_, torch.Tensor.true_divide_):
```
Comment (gchanan) [inline test/test_type_promotion.py:633]: would be nicer if you could just assert div_ is true_divide_ eh?

**Review by gchanan — APPROVED** [review]

File: test/test_type_promotion.py (hunk 11 above)
Comment (mruberry) [reply test/test_type_promotion.py:633]: It would be. We could still set the functions equal in Python and C++.

**Comment by mruberry** [comment]
@JackCaoG I was planning to land this on Monday, that OK with you? I can start the land in the morning and I would expect it to land before noon.

**Comment by JackCaoG** [comment]
> @JackCaoG I was planning to land this on Monday, that OK with you? I can start the land in the morning and I would expect it to land before noon.

Sounds good

**Comment by mruberry** [comment]
> > @JackCaoG I was planning to land this on Monday, that OK with you? I can start the land in the morning and I would expect it to land before noon.
> 
> Sounds good

Thanks @JackCaoG!

**Comment by mruberry** [comment]
@JackCaoG This is merged (finally). Sorry it's later than expected due to some internal test slowdowns today.

**Comment by JackCaoG** [comment]
@mruberry thanks for the reminder, I merged xla side change too.
